homeassistant/components/schedule/__init__.py

```diff
@@ -175,6 +175,10 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     )
     await storage_collection.async_load()
 
+    # Expose the storage collection so it can be interacted with programatically
+    # by other components (not only via. the websockets UI).
```

Comment:
Other components should not directly use these.

As a matter of fact, we shouldn't be crossing/mixing components. Can you provide a concrete use case where this is used?

../Frenck

Reply:
Sure.

Right now, I want to use a schedule helper to manage the schedule for my [Power Pet Door](https://github.com/corporategoth/ha-powerpetdoor), which is an integration I have written.

For now, I have created a [Schedule entity](https://github.com/corporategoth/ha-powerpetdoor/blob/main/custom_components/powerpetdoor/schedule.py) - which can read the current schedule from the power pet door, and update it on the device if I set one.

HOWEVER, while this does create an entity that shows up in the helper list, because it's not actually stored in the backing store, the UI associated with Schedules (which uses websockets) can't edit it.  There is no way for me to actually edit the schedule via. the UI of HA.

The Schedule UI / entity is actually perfect for my needs, it's clean, simple, and has a weekly schedule.  I would LOVE to re-use the schedule component.

HOWEVER, in order to re-use the Schedule component/entity (and it's UI), I need to be able to:
1. Listen for changes to the schedule entity, and use that to send an update to the Power Pet Door device.
2. When I read the schedule from the power pet door device, I need to be able to update/replace the schedule in the Schedule entity (which has to update the backing store), so that the UI via. websockets can see the updates and render them if I click on the entity.

The only way to do this right now would be some REALLY kludgy work trying to invoke the websockets API directly from my component, etc.  Or essentially copy/paste the schedule UI stuff and create my own websockets API (which is unnecessary and something I don't want to maintain).

The solution in this PR, while not perfect, allows me to re-write my custom schedule entity to use a standard schedule entity, then listen for changes on it, or invoke the update directly to the storage, which achieves my intended goal.

If Schedule was a 'first class' entity, like Switch, or Cover, with the associated UI changes to allow for editing a schedule (ie. the UI that exists for the helper Schedule entity) - as was proposed [here](https://community.home-assistant.io/t/new-schedule-entity-so-it-can-be-used-in-integrations/468072), this would not be as necessary.  I could create an instance of a Schedule and it would be editable just fine - with functions I could override.  Like any switch, cover, etc.  No problem.  But Schedule is not a 'first class' entity, so I'm trying to work around it, by being able to interact with the underlying storage directly.  Which only works if I have a reference to the underlying storage.

Reply:
If you have a better idea of how I could use a Schedule helper with my integration so that 1) the schedule edit UI works, 2) when a schedule is edited, I can be notified of the changes and send them to my device, and 3) when I read the schedule from my device it can update the schedule helper, then I'm all ears.

Reply:
> Right now, I want to use a schedule helper to manage the schedule for my [Power Pet Door](https://github.com/corporategoth/ha-powerpetdoor), which is an integration I have written.

This is not an entity component platform, this should not be used by, or be provided by,  other integrations. The provided use case isn't valid.

> If you have a better idea of how I could use a Schedule helper with my integration 

Yes: Don't. They are not designed for that purpose.

Reply:
Sure.  But it's very clear that you're never going to actually create a Schedule entity that IS fit for purpose (feature request https://community.home-assistant.io/t/new-schedule-entity-so-it-can-be-used-in-integrations/468072), that's not just a helper.  So what you're essentially saying is that schedules are unsupported by Home Assistant.

Sure, there is a helper, but that should ONLY be edited via. the UI.  And should ONLY be used to indicate if you are in or out of the scheduled times (show a value of on or off).  NOT for actually storing a schedule that can be used by anything else, or passed around as a full schedule, etc.

I mean, sure it's not what it was designed for, but if you're not going to make something that IS fit for purpose, what's the harm in giving people the ability to MAKE it work for them if they can?
In the words of NASA, "I don't care what it's designed to do, I want to know what it CAN do..."

ie. try the Linux model, enabling hackers to hack, not the Apple model, of locking everything down.

Reply:
I mean, they can just create a local calendar and put automations to turn something on or off. With recurring calendar items this is quite easy to do.

This approach will suit your basic users and advanced users which might want to take holidays in account or the location of the pet since they can edit their own automations to whatever they want.

Reply:
The problem is, **the device itself** has a schedule.  This is not a case of me wanting to create a schedule and have home assistant fire events at those scheduled times.

The schedule is stored IN THE HARDWARE device itself.  Once you send the schedule, the device itself turns things on or off based on that schedule.

So I need a way to edit said schedule **and send it to the device**, or read the schedule from the device **and update the home assistant entity**

The schedule in the hardware is a weekly schedule only (it only stores Monday - Sunday, 00:00 - 23:59).  Not a calendar.

If this was just a case of me wanting to send an event to the hardware at specific times, I could EASILY create a schedule, or calendar event and fire events off that in Home Assistant.  But that's not what I'm trying to achieve.  I'm trying to support the built-in schedule in the hardware.